### PR TITLE
feat(M): adding daily session limits for both user and admin

### PR DESCRIPTION
**Pull request by adamgdoura6**
## 💡 Proposed Solution

**1. Configuration (Backend)**
The daily session limit for users is stored in the database in the AppConfig table, under the key 'dailyUserSessionLimit'.
This value can be set and updated by an admin via the admin dashboard (frontend), which calls the backend /app-configs endpoint to update the value.

**2. Enforcement (Backend)**
When a user (non-admin) tries to create a new session:

- The backend (SessionService.create_new_session) fetches the current daily session limit from AppConfig.
- It counts how many sessions the user has started today (sessions_created_today).
- If the user has already reached the daily limit, the backend returns a 403 Forbidden error and does not allow the session to be created.
- If not, the session is created, and the user's sessions_created_today counter is incremented.

Tracking:
Each user has two fields in their profile:

- sessions_created_today: How many sessions they have started today.
- sessions_created_today_date: The date for which the above counter is valid (resets daily).

**3. Exposure in API Responses (Backend)**
The remaining number of sessions a user can start today (num_remaining_daily_sessions) is included in the response body of:

- GET /user-profile/profile
- GET /user-profile
- GET /user-profile/stats
- GET /user-profile/export

The daily session limit itself is also included in the user stats response.

**4. Admin Controls (Frontend)**
Admins can view and update the daily session limit from the admin dashboard.
The admin dashboard fetches the current limit and allows editing it via a dedicated UI component, which calls the backend to update the value.

**5. User Experience (Frontend)**

- Non-admin users see a stat card on their dashboard showing how many sessions they have left for the day (e.g., “2/3 sessions left today”). admin users see the stat card with only the session limit (e.g. #/3)
- This is fetched from the backend as part of the user stats/profile and displayed only for non-admins.
- If a user tries to start a session after reaching their limit, they will receive an error and cannot proceed.

**6. Export**
When a user exports their data, the export includes the num_remaining_daily_sessions field, so users can see their session usage at the time of export.

## ✅ Local Testing Instructions

Best way to test it is I believe, to run the app, connect with user and admin, and test edge cases. examples:

- no sessions left and user tries to create new one
- the number of sessions actually resets at the end of the day
- ..

## 🔬 Developer Checklist (must complete before marking PR as **Ready for Review**)

> Keep the PR as a **draft** until all of the following are checked and you're confident it's ready for review.

- [ ] I tested the app locally using `uv run fastapi dev` or equivalent.
- [ ] I ran `docker compose down -v` to removing all volumes and then `docker compose up --build` to build and start the app.
- [ ] I confirmed that all new or changed **environment variables** are:

  - [ ] The app has been tested with the new environment variables removed (e.g. deleted from .env, app started in a fresh terminal).
  - [ ] Documented clearly in `.env.example` or a config file.
  - [ ] Defaulted sensibly (or fail-fast if truly required).
  - [ ] Communicated to other stakeholders (if needed).

- [ ] All migrations or DB changes have been tested locally.
- [ ] I added or updated relevant unit/integration tests.
- [ ] I included clear **testing instructions** in this PR.
- [ ] I have considered performance, logging, and error handling.

## 🧠 Reviewer Notes

- Which parts of the code need extra attention?
- Is this a functional change, a refactor, or a cleanup?
- Any known issues or follow-up work?

### CODE REVIEW

File: backend/app/schemas/user_profile.py [hunk 1]
```diff
@@ -83,3 +86,6 @@ class UserStatisticsRead(CamelModel):
     goals_achieved: int  # summation of all goals achieved
     performance_over_time: list[int]
     skills_performance: dict[str, int]
+    daily_session_limit: int | None = None
+    remaining_sessions_today: int | None = None
```
Comment (ta-mey) [inline backend/app/schemas/user_profile.py:90]: The only value we need here is daily session limit right? Why include the two others? Same goes for other endpoints.

File: backend/app/services/admin_dashboard_service.py [hunk 2]
```diff
@@ -52,4 +61,5 @@ def get_admin_dashboard_stats(self) -> AdminDashboardStatsRead:
             total_reviews=total_reviews,
             score_sum=stats.score_sum,
             daily_token_limit=daily_token_limit,
```
Comment (ta-mey) [inline backend/app/services/admin_dashboard_service.py:63]: Please remove daily token limit.

File: backend/app/services/session_service.py [hunk 3]
```diff
@@ -102,6 +103,29 @@ def fetch_paginated_sessions(
...
+                # Count user's sessions started today
+                today = datetime.now(UTC).date()
+                scenario_ids = self._get_user_scenario_ids(user_profile.id)
+                if scenario_ids:
+                    session_count = self.db.exec(
+                        select(func.count())
+                        .select_from(Session)
+                        .where(Session.scenario_id.in_(scenario_ids))
+                        .where(func.date(Session.started_at) == today)
+                    ).one()
```
Comment (ta-mey) [inline backend/app/services/session_service.py:123]: You can't get the session count from the session table because users have the ability to delete sessions and therefore bypass this daily session limit. We need to have a session counter in the database that we increment with each new session. More details are in [the issue](https://github.com/jst-seminar-rostlab-tum/personio-foundation-coachai/issues/554).

File: backend/app/services/session_service.py [hunk 4]
```diff
@@ -117,6 +141,14 @@ def create_new_session(
         new_session.status = SessionStatus.started
         new_session.started_at = datetime.now(UTC)
 
+        # Update user's daily session counter
+        today = new_session.started_at.date()
+        if user_profile.sessions_created_today_date != today:
```
Comment (ta-mey) [inline backend/app/services/session_service.py:146]: This check should be included int he above session limit check to ensure that if a new day is started, the session counter resets and a new session is allowed.

File: frontend/src/app/[locale]/(main)/dashboard/page.tsx [hunk 5]
```diff
@@ -56,6 +56,15 @@ export default async function DashboardPage() {
         />
         <StatCard value={`${userStats.currentStreakDays}d`} label={t('userStats.currentStreak')} />
         <StatCard value={`${userStats.averageScore ?? 0}%`} label={tCommon('avgScore')} />
+        {/* Remaining sessions stat card for non-admin users */}
+        {userProfile.accountRole !== 'admin' &&
```
Comment (ta-mey) [inline frontend/src/app/[locale]/(main)/dashboard/page.tsx:60]: This should be visible as an admin as well to prevent confusion. Just replace the actual number of remaining sessions with N/A since admins do not have a limit.

File: backend/app/models/live_feedback_model.py [hunk 6]
```diff
@@ -8,7 +8,7 @@
 
 class LiveFeedback(CamelModel, table=True):
     id: UUID = Field(default_factory=uuid4, primary_key=True)
-    session_id: UUID = Field(foreign_key='session.id', primary_key=False)
+    session_id: UUID = Field(foreign_key='session.id', ondelete='CASCADE')
```
[the remark this answers is not in the document]
Comment (rautenrieb) [reply backend/app/models/live_feedback_model.py:11]: Let's keep this change! It should be set to `ondelete='CASCADE'`! I think it was an oversight of the Live Feedback integration.

File: backend/app/schemas/user_profile.py (hunk 1 above)
Comment (rautenrieb) [reply backend/app/schemas/user_profile.py:90]: We only need `daily_session_limit` and `daily_sessions_completed` for the frontend to display limit and and sessions completed.

File: backend/app/services/user_export_service.py [hunk 7]
```diff
@@ -23,6 +23,26 @@
 def _build_export_user_profile(user_profile: UserProfile) -> ExportUserProfile:
     """Build export user profile from UserProfile model."""
     try:
+        # Fetch daily session limit from AppConfig
+        from app.models.app_config import AppConfig
+
+        db_session = (
+            user_profile.__sa_session__ if hasattr(user_profile, '__sa_session__') else DBSession()
+        )
```
Comment (rautenrieb) [inline backend/app/services/user_export_service.py:31]: What is this?

File: backend/app/services/user_export_service.py [hunk 8]
```diff
@@ -23,6 +23,26 @@
 def _build_export_user_profile(user_profile: UserProfile) -> ExportUserProfile:
     """Build export user profile from UserProfile model."""
     try:
+        # Fetch daily session limit from AppConfig
+        from app.models.app_config import AppConfig
```
Comment (rautenrieb) [inline backend/app/services/user_export_service.py:27]: Avoid imports in functions.

File: backend/app/services/user_profile_service.py [hunk 9]
```diff
@@ -29,6 +29,17 @@ def __init__(self, db: DBSession) -> None:
         self.db = db
 
     def _get_detailed_user_profile_response(self, user: UserProfile) -> UserProfileExtendedRead:
+        # Fetch daily session limit from AppConfig
+        from app.models.app_config import AppConfig
```
Comment (rautenrieb) [inline backend/app/services/user_profile_service.py:33]: Avoid imports in method please

File: backend/app/services/user_profile_service.py [hunk 10]
```diff
@@ -49,9 +60,22 @@ def _get_detailed_user_profile_response(self, user: UserProfile) -> UserProfileE
             ],
             updated_at=user.updated_at,
             store_conversations=user.store_conversations,
+            sessions_created_today=user.sessions_created_today,
+            sessions_created_today_date=user.sessions_created_today_date,
+            num_remaining_daily_sessions=num_remaining_daily_sessions,
         )
 
     def _get_user_profile_response(self, user: UserProfile) -> UserProfileRead:
+        from app.models.app_config import AppConfig
```
Comment (rautenrieb) [inline backend/app/services/user_profile_service.py:69]: Avoid imports in method please

File: backend/app/services/user_profile_service.py [hunk 11]
```diff
@@ -126,7 +153,16 @@ def get_user_statistics(self, user_id: UUID) -> UserStatisticsRead:
                 status_code=status.HTTP_404_NOT_FOUND,
                 detail='User profile not found.',
             )
-
+        from app.models.app_config import AppConfig
```
Comment (rautenrieb) [inline backend/app/services/user_profile_service.py:156]: Avoid imports in method please

**Review by rautenrieb — CHANGES_REQUESTED** [review]
See additional comments for backend. Can't comment on frontend.

**Comment by adamgdoura6** [comment]
> We should also make some changes to handling a non-existing session limit:
> 
> 1. If the session limit is not available (e.g. set to a null-value), we should assume the session limit was hit. There won't be a scenario where we don't want to enforce a limit and it can be dangerous to build this into the code.
> 2. The number of remaining sessions should always be returned as an integer from the backend. The fronted can also assume this and shouldn't be hiding the session limit statistic in the dashboard if it doesn't get a value from the backend.

this is done

**Comment by adamgdoura6** [comment]
@rautenrieb not sure what you mean with daily_sessions_completed. My logic sends to the frontend daily_session_limit and num_remaining_daily_sessions. If we needed the third variable "number of sessions created" for that day, we can substract these two.

**Comment by adamgdoura6** [comment]
@rautenrieb @ta-mey please if possible a second review. The frontend can be handled in a second PR. I just implemented the necessary items to test my solution

File: backend/app/schemas/user_profile.py (hunk 1 above)
Comment (ta-mey) [reply backend/app/schemas/user_profile.py]: The plan is to only display the number of remaining sessions for that user for the current day. For that we only need `num_remaining_daily_sessions`.

Comment (ta-mey) [reply backend/app/schemas/user_profile.py]: It is the only value that should be returned by any of the endpoints.

**Comment by ta-mey** [comment]
@adamgdoura6, please increment the session counter in the `/realtime-session/{session_id}` endpoint and not the POST `/sessions` endpoint. This is the one that matters, since it gives the user access to an authenticated OpenAI session.

The point of the session limit feature is to limit the cost that one user can incur in a day, and they would be able to do infinite number of conversations for the same session using `/realtime-session/{session_id}` as long as they never completed the session using PUT `/sessions/{id}`.

File: backend/app/database.py [hunk 12]
```diff
@@ -25,8 +25,7 @@
     else:
         SQLALCHEMY_DATABASE_URL = settings.database_url
 else:
-    SQLALCHEMY_DATABASE_URL = 'postgresql://postgres:postgres@127.0.0.1:54322/postgres'
-
+    SQLALCHEMY_DATABASE_URL = f'postgresql+psycopg://{settings.postgres_user}:{settings.postgres_password}@{settings.postgres_host}:{settings.postgres_port}/{settings.postgres_db}'
```
[the remark this answers is not in the document]
Comment (ta-mey) [reply backend/app/database.py:28]: I would avoid making changes like these, because it requires everyone in the team that runs the backend, which includes the frontend team now due to the way we set up local supabase, to make changes to their env file.

File: backend/app/config.py [hunk 13]
```diff
@@ -49,7 +49,7 @@ class Settings(BaseSettings):
     DEFAULT_CHEAP_MODEL: str = 'gemini-2.0-flash-lite-001'
     DEFAULT_MODEL: str = 'gemini-2.5-pro'
 
-    DEV_MODE_SKIP_AUTH: bool = True
+    DEV_MODE_SKIP_AUTH: bool = False
```
[the remark this answers is not in the document]
Comment (ta-mey) [reply backend/app/config.py:52]: This changes the behavior of the backend in a way that others will not expect, which can lead to confusion. I would avoid changing the default value here unless if there is a strong reason to do this since it lets us avoid any friction with aligning everyone on this change.

File: backend/app/services/session_service.py [hunk 14]
```diff
@@ -105,6 +106,34 @@ def fetch_paginated_sessions(
...
+                    detail='Daily session limit is not configured. '
+                    'Please contact an administrator.',
+                )
+
+            session_limit = int(session_limit_config)
+            # Update user's daily session counter
+            today = datetime.now(UTC).date()
+            if user_profile.last_session_date != today:
+                user_profile.sessions_created_today = 0
+                user_profile.last_session_date = today
```
Comment (ta-mey) [inline backend/app/services/session_service.py:129]: For consistency, let's reset the session limit for admin accounts as well.

**Review by ta-mey — CHANGES_REQUESTED** [review]
Looks great! A few minor changes still.

**Comment by adamgdoura6** [comment]
@ta-mey as i mentionned, i wont push the DEV_MODE_SKIP_AUTH and link to dev. its just for this branch and will delete these changements once everything else here is confirmed

**Comment by adamgdoura6** [comment]
@ta-mey made some changes please have a look

**Comment by ta-mey** [comment]
Ah, must have missed you mentioning that somewhere.

**Review by ta-mey — DISMISSED** [review]
LGTM, great job! If you tested locally and it works as expected, please go ahead and merge. Just don't forget to remove unnecessary changes beforehand.

File: frontend/src/services/AdminService.ts [hunk 15]
```diff
@@ -24,7 +24,22 @@ const updateDailyUserTokenLimit = async (api: AxiosInstance, limit: number) => {
...
+    return data;
+  } catch (error) {
+    console.error('Error updating daily user session limit: ', error);
+    throw error;
+  }
+};
+
 export const adminService = {
   getAdminStats,
   updateDailyUserTokenLimit,
```
Comment (JCaesar20) [inline frontend/src/services/AdminService.ts:43]: do we still need the updateDailyUserTokenLimit if no please remove ?

File: frontend/src/app/[locale]/(main)/dashboard/page.tsx [hunk 16]
```diff
@@ -56,6 +56,15 @@ export default async function DashboardPage() {
         />
         <StatCard value={`${userStats.currentStreakDays}d`} label={t('userStats.currentStreak')} />
         <StatCard value={`${userStats.averageScore ?? 0}%`} label={tCommon('avgScore')} />
+        {/* Remaining sessions stat card - always shown */}
+        <StatCard
+          value={
+            userProfile.accountRole === 'admin'
```
Comment (JCaesar20) [inline frontend/src/app/[locale]/(main)/dashboard/page.tsx:62]: do the check please before the return, like const isAdmin = userProfile.accountRole === 'admin'; and use isAdmin here directly

File: frontend/src/app/[locale]/(main)/admin/page.tsx [hunk 17]
```diff
@@ -50,7 +50,7 @@ export default async function AdminPage() {
             <StatCard key={i} value={stat.value} label={stat.label} />
           ))}
         </div>
-        <TokenSetter dailyTokenLimit={stats.dailyTokenLimit} />
```
Comment (JCaesar20) [inline frontend/src/app/[locale]/(main)/admin/page.tsx:53]: Please delete TokenSetter component files

File: frontend/messages/en/Dashboard.json [hunk 18]
```diff
@@ -7,7 +7,8 @@
   },
   "userStats": {
     "trainingTime": "Training Time",
-    "currentStreak": "Current Streak"
+    "currentStreak": "Current Streak",
+    "remainingSessionsToday": "Remaining sessions today"
```
Comment (JCaesar20) [inline frontend/messages/en/Dashboard.json:11]: the new translation keys are missing in de/Dashboard.json

File: frontend/messages/en/Admin.json [hunk 19]
```diff
@@ -20,5 +20,9 @@
   "sortNewest": "Newest",
   "sortOldest": "Oldest",
   "sortHighest": "Highest",
-  "sortLowest": "Lowest"
+  "sortLowest": "Lowest",
+  "sessionsPerUserLabel": "Sessions per user per day",
+  "sessionSuccess": "Session limit saved successfully",
+  "sessionFailed": "Error saving session limit",
+  "sessionNumberFailed": "Please enter a valid number (1-1000)"
```
Comment (JCaesar20) [inline frontend/messages/en/Admin.json:27]: the new translation keys are missing in de/Admin.json

Comment (JCaesar20) [reply frontend/messages/en/Admin.json:27]: Please delete these from en and de  if you wont use them
"tokenSuccess": "Token limit saved successfully",
"tokenFailed": "Error saving token limit",
"tokenNumberFailed": "Please enter a valid number (1-1000000)",
"tokensPerUserLabel": "Sessions/User/Day",

**Comment by adamgdoura6** [comment]
@ta-mey  @JCaesar20  all comments regarding the backend are resolved. if you noticed anything else, tag me or text in slack
